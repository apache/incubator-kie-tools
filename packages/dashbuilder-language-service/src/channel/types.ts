/*
 * Copyright 2022 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { TextDocument } from "vscode-languageserver-textdocument";
import { CompletionItemKind, Range } from "vscode-languageserver-types";

export declare type DashbuilderLsNodeType = "object" | "array" | "property" | "string" | "number" | "boolean" | "null";
export declare type DashbuilderJsonPath = (string | number)[];

/**
 * The AST node used in the LanguageServices
 */
export type DashbuilderLsNode = {
  type: DashbuilderLsNodeType;
  value?: any;
  offset: number;
  length: number;
  colonOffset?: number;
  parent?: DashbuilderLsNode;
  children?: DashbuilderLsNode[];
};

export interface TranslateArgs {
  completion: object | string;
  completionItemKind: CompletionItemKind;
  currentNodeRange?: Range;
  cursorOffset: number;
  document: TextDocument;
  overwriteRange?: Range;
}
