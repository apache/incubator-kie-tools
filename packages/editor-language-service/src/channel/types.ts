/*
 * Copyright 2023 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Position, TextDocument } from "vscode-languageserver-textdocument";
import { CompletionItemKind, Range } from "vscode-languageserver-types";
import { EditorLanguageServiceCommandTypes } from "../api";

// types ElsJsonPath, ELsNode, ELsNodeType need to be compatible with jsonc types
export declare type ElsJsonPath = (string | number)[];
export declare type ELsNodeType = "object" | "array" | "property" | "string" | "number" | "boolean" | "null";

type OmitDistributive<T, K extends PropertyKey> = T extends any
  ? T extends object
    ? Id<OmitRecursively<T, K>>
    : T
  : never;
type Id<T> = {} & { [P in keyof T]: T[P] };
export type OmitRecursively<T extends any, K extends PropertyKey> = Omit<
  { [P in keyof T]: OmitDistributive<T[P], K> },
  K
>;

/**
 * The AST node used in the LanguageServices
 */
export type ELsNode = {
  type: ELsNodeType;
  value?: any;
  offset: number;
  length: number;
  colonOffset?: number;
  parent?: ELsNode;
  children?: ELsNode[];
};

export interface ShouldCompleteArgs {
  content: string;
  cursorOffset: number;
  cursorPosition: Position;
  node: ELsNode | undefined;
  path: ElsJsonPath;
  root: ELsNode | undefined;
}

export interface ShouldCreateCodelensArgs {
  content: string;
  node: ELsNode;
  commandName: EditorLanguageServiceCommandTypes;
}

export interface TranslateArgs {
  completion: object | string;
  completionItemKind: CompletionItemKind;
  currentNodeRange?: Range;
  cursorOffset: number;
  document: TextDocument;
  overwriteRange?: Range;
}

export interface CodeCompletionStrategy {
  translate(args: TranslateArgs): string;
  formatLabel(label: string, completionItemKind: CompletionItemKind): string;
  shouldComplete(args: ShouldCompleteArgs): boolean;
  getStartNodeValuePosition(document: TextDocument, node: ELsNode): Position | undefined;
  shouldCreateCodelens(args: ShouldCreateCodelensArgs): boolean;
}
