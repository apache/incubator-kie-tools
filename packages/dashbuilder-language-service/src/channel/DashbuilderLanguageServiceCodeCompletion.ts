/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {
  EditorLanguageServiceCodeCompletionFunctions,
  EditorLanguageServiceCodeCompletionFunctionsArgs,
  EditorLanguageServiceEmptyFileCodeCompletionFunctionArgs,
} from "@kie-tools/json-yaml-language-service/dist/channel";
import { CompletionItem, CompletionItemKind, InsertTextFormat, Range } from "vscode-languageserver-types";
import { dashbuilderCompletion } from "../assets/code-completions/";

export type DashbuilderLanguageServiceCodeCompletionFunctionsArgs =
  EditorLanguageServiceCodeCompletionFunctionsArgs & {};

/**
 * DashbuilderLanguageService CodeCompletion functions
 */
export const DashbuilderLanguageServiceCodeCompletion: EditorLanguageServiceCodeCompletionFunctions = {
  getEmptyFileCodeCompletions(
    args: EditorLanguageServiceEmptyFileCodeCompletionFunctionArgs
  ): Promise<CompletionItem[]> {
    const kind = CompletionItemKind.Text;
    const label = "Create your first dashboard";
    return Promise.resolve([
      {
        kind,
        label,
        detail: "Start with a simple dashboard",
        sortText: `100_${label}`, //place the completion on top in the menu
        textEdit: {
          newText: args.codeCompletionStrategy.translate({
            ...args,
            completion: dashbuilderCompletion,
            completionItemKind: kind,
          }),
          range: Range.create(args.cursorPosition, args.cursorPosition),
        },
        insertTextFormat: InsertTextFormat.Snippet,
      },
    ]);
  },
};
