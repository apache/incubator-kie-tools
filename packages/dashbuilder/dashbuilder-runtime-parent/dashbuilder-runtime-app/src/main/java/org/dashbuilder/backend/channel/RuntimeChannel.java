/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.dashbuilder.backend.channel;

import javax.enterprise.context.ApplicationScoped;
import javax.enterprise.event.Observes;
import javax.inject.Inject;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.sse.OutboundSseEvent;
import javax.ws.rs.sse.Sse;
import javax.ws.rs.sse.SseBroadcaster;
import javax.ws.rs.sse.SseEventSink;

import org.dashbuilder.backend.RuntimeOptions;
import org.dashbuilder.shared.event.RemovedRuntimeModelEvent;
import org.dashbuilder.shared.event.SSEType;
import org.dashbuilder.shared.event.UpdatedRuntimeModelEvent;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Channel with client for runtime events.
 *
 */
@Path("runtime-channel")
@ApplicationScoped
public class RuntimeChannel {

    Logger logger = LoggerFactory.getLogger(RuntimeChannel.class);

    private Sse sse;
    private SseBroadcaster sseBroadcaster;
    private OutboundSseEvent.Builder eventBuilder;

    @Inject
    RuntimeOptions runtimeOptions;


    @GET
    @Path("subscribe")
    @Produces(MediaType.SERVER_SENT_EVENTS)
    public void listen(@Context SseEventSink sseEventSink) {
        if (runtimeOptions.isWatchModels() && sseBroadcaster != null) {
            sseBroadcaster.register(sseEventSink);
            sseEventSink.send(sse.newEvent(SSEType.SUBSCRIBED.name(), ""));
        } else {
            sseEventSink.send(sse.newEvent(SSEType.NOT_SUBSCRIBED.name(), ""));
        }
    }

    public void onRuntimeModelUpdated(@Observes UpdatedRuntimeModelEvent updatedRuntimeModel) {
        broadcastEvent(SSEType.MODEL_UPDATED, updatedRuntimeModel.getRuntimeModelId());
    }

    public void onRuntimeModelRemoved(@Observes RemovedRuntimeModelEvent removedRuntimeModel) {
        broadcastEvent(SSEType.MODEL_REMOVED, removedRuntimeModel.getRuntimeModelId());
    }

    private void broadcastEvent(SSEType type, String data) {
        if (sseBroadcaster != null) {
            var sseEvent = eventBuilder.name(type.name())
                                       .mediaType(MediaType.TEXT_PLAIN_TYPE)
                                       .data(data)
                                       .reconnectDelay(3000)
                                       .build();
            sseBroadcaster.broadcast(sseEvent);
        }
    }

    @Context
    public void setSse(Sse sse) {
        this.sse = sse;
        this.eventBuilder = sse.newEventBuilder();
        this.sseBroadcaster = sse.newBroadcaster();
    }

}
