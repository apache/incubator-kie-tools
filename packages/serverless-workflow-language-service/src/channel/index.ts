export * from "./SwfJsonLanguageService";
export * from "./SwfLanguageService";
export * from "./SwfYamlLanguageService";
