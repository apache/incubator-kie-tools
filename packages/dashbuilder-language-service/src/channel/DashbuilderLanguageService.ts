/*
 * Copyright 2022 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { TextDocument } from "vscode-languageserver-textdocument";
import { CodeLens, CompletionItem, Diagnostic, DiagnosticSeverity, Position, Range } from "vscode-languageserver-types";
import { FileLanguage } from "../api";
import * as jsonc from "jsonc-parser";
import { DashbuilderLanguageServiceCodeCompletion } from "./DashbuilderLanguageServiceCodeCompletion";
import { DashbuilderLanguageServiceCodeLenses } from "./DashbuilderLanguageServiceCodeLenses";
import { DashbuilderJsonPath, DashbuilderLsNode } from "./types";
import {
  Kind,
  load,
  YAMLAnchorReference,
  YamlMap,
  YAMLMapping,
  YAMLNode,
  YAMLScalar,
  YAMLSequence,
} from "yaml-language-server-parser";
import {
  getLanguageService,
  LanguageSettings,
  SchemaRequestService,
  SettingsState,
  Telemetry,
  WorkspaceContextService,
} from "@kie-tools/yaml-language-server";
import { Connection } from "vscode-languageserver/node";
import { DASHBUILDER_SCHEMA } from "../assets/schemas";

export class DashbuilderLanguageService {
  constructor() {}

  parseContent(content: string): DashbuilderLsNode | undefined {
    if (!content.trim()) {
      return;
    }

    const ast = load(content);

    // check if the yaml is not valid
    if (ast && ast.errors && ast.errors.length) {
      return;
    }

    return astConvert(ast);
  }

  public async getCompletionItems(args: {
    content: string;
    uri: string;
    cursorPosition: Position;
    cursorWordRange: Range;
  }): Promise<CompletionItem[]> {
    const rootNode = this.parseContent(args.content);
    const doc = TextDocument.create(args.uri, FileLanguage.YAML, 0, args.content);
    const cursorOffset = doc.offsetAt(args.cursorPosition);

    if (
      args.content.slice(cursorOffset - 1, cursorOffset) === ":" ||
      args.content.slice(cursorOffset - 1, cursorOffset) === "-"
    ) {
      return [];
    }

    const isCurrentNodeUncompleted = rootNode
      ? isNodeUncompleted({
          ...args,
          rootNode,
          cursorOffset,
        })
      : false;

    if (isCurrentNodeUncompleted) {
      args.cursorPosition = Position.create(args.cursorPosition.line, args.cursorPosition.character - 1);
    }

    if (!rootNode) {
      return args.content.trim().length
        ? []
        : DashbuilderLanguageServiceCodeCompletion.getEmptyFileCodeCompletions({
            ...args,
            cursorOffset,
            document: doc,
          });
    }
    return [];
  }

  public async getCodeLenses(args: { content: string; uri: string }): Promise<CodeLens[]> {
    const rootNode = this.parseContent(args.content);
    if (!args.content.trim().length) {
      return DashbuilderLanguageServiceCodeLenses.createNewDashboard();
    }

    if (!rootNode) {
      return [];
    }
    return [];
  }

  public async getDiagnostics(args: { content: string; uriPath: string }): Promise<Diagnostic[]> {
    if (!args.content.trim()) {
      return [];
    }
    const rootNode = this.parseContent(args.content);
    const loadErrors = !rootNode ? load(args.content).errors : [];
    //check the syntax
    if (loadErrors.length > 0) {
      const error = loadErrors[0];
      const position = Position.create(error.mark.line, error.mark.column);
      // show only the first error because syntax errors are repeated for each line, after the error.
      return [
        {
          message: error.message,
          range: Range.create(position, position),
          severity: DiagnosticSeverity.Error,
        },
      ];
    }

    if (!rootNode) {
      return [];
    }

    // this ensure the document is validated again
    const docVersion = Math.floor(Math.random() * 1000);

    const textDocument = TextDocument.create(
      args.uriPath,
      `dashbuilder-${FileLanguage.YAML}`,
      docVersion,
      args.content
    );

    return await this.getSchemaDiagnostics(textDocument, ["*.dash.yaml", "*.dash.yml", "*.dash.json"]);
  }

  private async getSchemaDiagnostics(textDocument: TextDocument, fileMatch: string[]): Promise<Diagnostic[]> {
    const schemaRequestService: SchemaRequestService = async (uri: string) => {
      if (uri === DASHBUILDER_SCHEMA.$id) {
        return Promise.resolve(JSON.stringify(DASHBUILDER_SCHEMA));
      } else {
        throw new Error(`Unable to load schema from '${uri}'`);
      }
    };
    const workspaceContext: WorkspaceContextService = {
      resolveRelativePath: (_relativePath: string, _resource: string) => {
        return "";
      },
    };

    const connection = {} as Connection;
    connection.onRequest = () => null;
    const telemetry = new Telemetry(connection);

    const yamlSettings = { yamlFormatterSettings: { enable: false } } as SettingsState;
    const yamlLanguageSettings: LanguageSettings = {
      validate: true,
      completion: true,
      format: false,
      hover: false,
      isKubernetes: false,
      schemas: [{ fileMatch, uri: DASHBUILDER_SCHEMA.$id }],
    };
    const yamlLs = getLanguageService(schemaRequestService, workspaceContext, connection, telemetry, yamlSettings);
    yamlLs.configure(yamlLanguageSettings);
    return yamlLs.doValidation(textDocument, false);
  }

  public dispose() {
    // empty for now
  }
}

const astConvert = (node: YAMLNode, parentNode?: DashbuilderLsNode): DashbuilderLsNode => {
  const convertedNode: DashbuilderLsNode = {
    type: "object",
    offset: node.startPosition,
    length: node.endPosition - node.startPosition,
    parent: parentNode,
  };

  if (node.kind === Kind.SCALAR) {
    convertedNode.value = (node as YAMLScalar).value;
    convertedNode.type = "string";
  } else if (node.kind === Kind.MAP) {
    const yamlMap = node as YamlMap;
    convertedNode.value = yamlMap.value;
    convertedNode.children = yamlMap.mappings.map((mapping) => astConvert(mapping, convertedNode));
    convertedNode.type = "object";
  } else if (node.kind === Kind.MAPPING) {
    const yamlMapping = node as YAMLMapping;
    convertedNode.value = yamlMapping.value;
    convertedNode.children = [
      astConvert(yamlMapping.key, convertedNode),
      ...(convertedNode.value ? [astConvert(yamlMapping.value, convertedNode)] : []),
    ];
    convertedNode.type = "property";
    convertedNode.colonOffset = yamlMapping.key.endPosition;
  } else if (node.kind === Kind.SEQ) {
    convertedNode.children = (node as YAMLSequence).items
      .filter((item) => item)
      .map((item) => astConvert(item, convertedNode));
    convertedNode.type = "array";
  } else if (node.kind === Kind.ANCHOR_REF || node.kind === Kind.INCLUDE_REF) {
    convertedNode.value = (node as YAMLAnchorReference).value;
    convertedNode.type = "object";
  }

  return convertedNode;
};

export function findNodeAtOffset(
  root: DashbuilderLsNode,
  offset: number,
  includeRightBound?: boolean
): DashbuilderLsNode | undefined {
  return jsonc.findNodeAtOffset(root as jsonc.Node, offset, includeRightBound) as DashbuilderLsNode;
}

export function getNodePath(node: DashbuilderLsNode): DashbuilderJsonPath {
  return jsonc.getNodePath(node as jsonc.Node);
}

export const isNodeUncompleted = (args: {
  content: string;
  uri: string;
  rootNode: DashbuilderLsNode;
  cursorOffset: number;
}): boolean => {
  if (args.content.slice(args.cursorOffset - 1, args.cursorOffset) !== " ") {
    return false;
  }

  const nodeAtPrevOffset = findNodeAtOffset(args.rootNode, args.cursorOffset - 1, true);

  if (!nodeAtPrevOffset) {
    return false;
  }

  return nodeAtPrevOffset.offset + nodeAtPrevOffset.length === args.cursorOffset - 1;
};

export const positions_equals = (a: Position | null, b: Position | null): boolean =>
  a?.line === b?.line && a?.character == b?.character;
