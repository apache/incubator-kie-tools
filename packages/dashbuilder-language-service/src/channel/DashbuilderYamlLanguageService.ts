/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {
  EditorYamlCodeCompletionStrategy,
  EditorYamlLanguageService,
  ELsNode,
  ELsShouldCreateCodelensArgs,
  IEditorLanguageService,
  indentText,
  positions_equals,
  ShouldCompleteArgs,
  TranslateArgs,
} from "@kie-tools/json-yaml-language-service/dist/channel";
import { TextDocument } from "vscode-languageserver-textdocument";
import { CodeLens, CompletionItem, CompletionItemKind, Diagnostic, Position, Range } from "vscode-languageserver-types";
import { dump } from "yaml-language-server-parser";
import { FileLanguage } from "../api";
import { DASHBUILDER_SCHEMA } from "../assets/schemas";
import { DashbuilderLanguageService } from "./DashbuilderLanguageService";
import { CodeCompletionStrategy, ShouldCreateCodelensArgs } from "./types";

export class DashbuilderYamlLanguageService implements IEditorLanguageService {
  private readonly ls: DashbuilderLanguageService;
  private readonly yamlELs: EditorYamlLanguageService;
  private readonly codeCompletionStrategy: DashbuilderYamlCodeCompletionStrategy;

  constructor() {
    this.ls = new DashbuilderLanguageService({
      fs: {},
      lang: {
        fileLanguage: FileLanguage.YAML,
        fileMatch: ["*.dash.yaml", "*.dash.yml"],
      },
    });
    this.codeCompletionStrategy = new DashbuilderYamlCodeCompletionStrategy();
    this.yamlELs = new EditorYamlLanguageService({
      ls: this.ls,
      codeCompletionStrategy: this.codeCompletionStrategy,
    });
  }

  public async getCompletionItems(args: {
    content: string;
    uri: string;
    cursorPosition: Position;
    cursorWordRange: Range;
  }): Promise<CompletionItem[]> {
    return await this.yamlELs.getCompletionItems(args);
  }

  public async getCodeLenses(args: { content: string; uri: string }): Promise<CodeLens[]> {
    return await this.yamlELs.getCodeLenses(args);
  }

  public async getDiagnostics(args: { content: string; uriPath: string }): Promise<Diagnostic[]> {
    return await this.yamlELs.getDiagnostics({
      ...args,
      getSchemaDiagnostics: (args: { textDocument: TextDocument; fileMatch: string[] }) =>
        this.getSchemaDiagnostics(args),
    });
  }

  public async getSchemaDiagnostics(args: { textDocument: TextDocument; fileMatch: string[] }): Promise<Diagnostic[]> {
    return await this.yamlELs.getSchemaDiagnostics({
      ...args,
      jsonSchema: DASHBUILDER_SCHEMA,
    });
  }

  public dispose() {
    return this.yamlELs.dispose();
  }
}

export class DashbuilderYamlCodeCompletionStrategy implements CodeCompletionStrategy {
  private eLsCodeCompletionStrategy: EditorYamlCodeCompletionStrategy;

  constructor() {
    this.eLsCodeCompletionStrategy = new EditorYamlCodeCompletionStrategy();
  }

  public translate(args: TranslateArgs): string {
    const completionDump = dump(args.completion, {}).slice(2, -1).trim();
    if (["{}", "[]"].includes(completionDump) || args.completionItemKind === CompletionItemKind.Text) {
      return completionDump;
    }
    const skipFirstLineIndent = args.completionItemKind !== CompletionItemKind.Module;
    const completionItemNewLine = args.completionItemKind === CompletionItemKind.Module ? "\n" : "";
    const completionText = completionItemNewLine + indentText(completionDump, 2, " ", skipFirstLineIndent);
    return ([CompletionItemKind.Interface, CompletionItemKind.Reference] as CompletionItemKind[]).includes(
      args.completionItemKind
    ) && positions_equals(args.overwriteRange?.start ?? null, args.currentNodeRange?.start ?? null)
      ? `- ${completionText}\n`
      : completionText;
  }

  public formatLabel(label: string, completionItemKind: CompletionItemKind): string {
    return this.eLsCodeCompletionStrategy.formatLabel(label, completionItemKind);
  }

  public getStartNodeValuePosition(document: TextDocument, node: ELsNode): Position | undefined {
    return this.eLsCodeCompletionStrategy.getStartNodeValuePosition(document, node);
  }

  public shouldComplete(args: ShouldCompleteArgs): boolean {
    return this.eLsCodeCompletionStrategy.shouldComplete(args);
  }

  public shouldCreateCodelens(args: ShouldCreateCodelensArgs): boolean {
    return this.eLsCodeCompletionStrategy.shouldCreateCodelens(args as ELsShouldCreateCodelensArgs);
  }
}
