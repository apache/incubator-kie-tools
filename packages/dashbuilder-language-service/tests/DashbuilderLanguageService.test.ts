/*
 * Copyright 2022 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { CodeLens, Position } from "vscode-languageserver-types";
import { DashbuilderLanguageService } from "@kie-tools/dashbuilder-language-service/dist/channel";
import { codeCompletionTester, ContentWithCursor } from "./testUtils";

const documentUri = "test.sw.yaml";

describe("Code lenses", () => {
  describe("empty file code lenses", () => {
    test.each([
      ["empty object / using JSON format", `{}`],
      ["empty object / with cursor before the object / using JSON format", `{}`],
      ["empty object / with cursor after the object / using JSON format", `{}`],
    ])("%s", async (_description, content: ContentWithCursor) => {
      const ls = new DashbuilderLanguageService();

      const codeLenses = await ls.getCodeLenses({ uri: documentUri, content });

      expect(codeLenses).toHaveLength(0);
    });
  });

  test.each([
    ["total empty file", ``],
    ["empty file with a newline before the cursor", `\n`],
    ["empty file with a newline after the cursor", `\n`],
  ])("%s", async (_description, content: ContentWithCursor) => {
    const ls = new DashbuilderLanguageService();

    const codeLenses = await ls.getCodeLenses({ uri: documentUri, content });
    const position = Position.create(0, 0);

    expect(codeLenses).toHaveLength(1);
    expect(codeLenses[0]).toStrictEqual({
      range: { start: position, end: position },
      command: {
        title: "Create a dashboard",
        command: "dashbuilder.ls.commands.OpenCompletionItems",
        arguments: [{ newCursorPosition: position }],
      },
    } as CodeLens);
  });
});

describe("code completion", () => {
  const ls = new DashbuilderLanguageService();
  describe("empty file completion", () => {
    test.each([
      ["empty object / using JSON format", `{ðŸŽ¯}`],
      ["empty object / with cursor before the object / using JSON format", `ðŸŽ¯{}`],
      ["empty object / with cursor after the object / using JSON format", `{}ðŸŽ¯`],
    ])("%s", async (_description, content: ContentWithCursor) => {
      const { completionItems } = await codeCompletionTester(ls, documentUri, content, false);

      expect(completionItems).toHaveLength(0);
    });

    test.each([
      ["total empty file", `ðŸŽ¯`],
      ["empty file with a newline before the cursor", `\nðŸŽ¯`],
      ["empty file with a newline after the cursor", `ðŸŽ¯\n`],
    ])("%s", async (_description, content: ContentWithCursor) => {
      const { completionItems } = await codeCompletionTester(ls, documentUri, content, false);

      expect(completionItems.length).toMatchSnapshot();
      expect(completionItems).toMatchSnapshot();
    });
  });
});
