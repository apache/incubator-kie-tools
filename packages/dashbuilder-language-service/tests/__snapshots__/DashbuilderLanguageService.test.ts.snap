// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code completion empty file completion empty file with a newline after the cursor 1`] = `1`;

exports[`code completion empty file completion empty file with a newline after the cursor 2`] = `
Array [
  Object {
    "detail": "Start with a simple dashboard",
    "insertTextFormat": 2,
    "kind": 1,
    "label": "Create your first dashboard",
    "sortText": "100_Create your first dashboard",
    "textEdit": Object {
      "newText": "datasets:
- uuid: products
  content: >-
            [
              [\\"Computers\\", \\"Scanner\\", 5, 3],
              [\\"Computers\\", \\"Printer\\", 7, 4],
              [\\"Computers\\", \\"Laptop\\", 3, 2],
              [\\"Electronics\\", \\"Camera\\", 10, 7],
              [\\"Electronics\\", \\"Headphones\\", 5, 9]
            ]
  columns:
    - id: Section
      type: LABEL
    - id: Product
      type: LABEL
    - id: Quantity
      type: NUMBER
    - id: Quantity2
      type: NUMBER
pages:
- components:
    - html: Welcome to Dashbuilder!
      properties:
        font-size: xx-large
        margin-bottom: 30px
    - settings:
        type: BARCHART
        lookup:
            uuid: products
            group:
                - columnGroup:
                    source: Product
                  functions:
                    - source: Product
                    - source: Quantity
                      function: SUM
                    - source: Quantity2
                      function: SUM
    - settings:
        lookup:
            uuid: products",
      "range": Object {
        "end": Object {
          "character": 0,
          "line": 0,
        },
        "start": Object {
          "character": 0,
          "line": 0,
        },
      },
    },
  },
]
`;

exports[`code completion empty file completion empty file with a newline before the cursor 1`] = `1`;

exports[`code completion empty file completion empty file with a newline before the cursor 2`] = `
Array [
  Object {
    "detail": "Start with a simple dashboard",
    "insertTextFormat": 2,
    "kind": 1,
    "label": "Create your first dashboard",
    "sortText": "100_Create your first dashboard",
    "textEdit": Object {
      "newText": "datasets:
- uuid: products
  content: >-
            [
              [\\"Computers\\", \\"Scanner\\", 5, 3],
              [\\"Computers\\", \\"Printer\\", 7, 4],
              [\\"Computers\\", \\"Laptop\\", 3, 2],
              [\\"Electronics\\", \\"Camera\\", 10, 7],
              [\\"Electronics\\", \\"Headphones\\", 5, 9]
            ]
  columns:
    - id: Section
      type: LABEL
    - id: Product
      type: LABEL
    - id: Quantity
      type: NUMBER
    - id: Quantity2
      type: NUMBER
pages:
- components:
    - html: Welcome to Dashbuilder!
      properties:
        font-size: xx-large
        margin-bottom: 30px
    - settings:
        type: BARCHART
        lookup:
            uuid: products
            group:
                - columnGroup:
                    source: Product
                  functions:
                    - source: Product
                    - source: Quantity
                      function: SUM
                    - source: Quantity2
                      function: SUM
    - settings:
        lookup:
            uuid: products",
      "range": Object {
        "end": Object {
          "character": 0,
          "line": 1,
        },
        "start": Object {
          "character": 0,
          "line": 1,
        },
      },
    },
  },
]
`;

exports[`code completion empty file completion total empty file 1`] = `1`;

exports[`code completion empty file completion total empty file 2`] = `
Array [
  Object {
    "detail": "Start with a simple dashboard",
    "insertTextFormat": 2,
    "kind": 1,
    "label": "Create your first dashboard",
    "sortText": "100_Create your first dashboard",
    "textEdit": Object {
      "newText": "datasets:
- uuid: products
  content: >-
            [
              [\\"Computers\\", \\"Scanner\\", 5, 3],
              [\\"Computers\\", \\"Printer\\", 7, 4],
              [\\"Computers\\", \\"Laptop\\", 3, 2],
              [\\"Electronics\\", \\"Camera\\", 10, 7],
              [\\"Electronics\\", \\"Headphones\\", 5, 9]
            ]
  columns:
    - id: Section
      type: LABEL
    - id: Product
      type: LABEL
    - id: Quantity
      type: NUMBER
    - id: Quantity2
      type: NUMBER
pages:
- components:
    - html: Welcome to Dashbuilder!
      properties:
        font-size: xx-large
        margin-bottom: 30px
    - settings:
        type: BARCHART
        lookup:
            uuid: products
            group:
                - columnGroup:
                    source: Product
                  functions:
                    - source: Product
                    - source: Quantity
                      function: SUM
                    - source: Quantity2
                      function: SUM
    - settings:
        lookup:
            uuid: products",
      "range": Object {
        "end": Object {
          "character": 0,
          "line": 0,
        },
        "start": Object {
          "character": 0,
          "line": 0,
        },
      },
    },
  },
]
`;

exports[`diagnostic empty file 1`] = `0`;

exports[`diagnostic empty file 2`] = `Array []`;

exports[`diagnostic valid 1`] = `0`;

exports[`diagnostic valid 2`] = `Array []`;
